define-test+run, remove-all-tests-in-package

some-mthfka opened the pull request:
Hi, could you please consider merging this?

Shinmera commented on report.lisp:
```diff
@@ -92,7 +92,9 @@
-            (setf (status result) :failed)))
+            (setf (status result) :failed)
+            (unless *silence-compilation-errors-p*
+              (error err))))
```
This re-signals errors, which destroys stack trace information. This needs to be reformulated using `handler-bind` instead to avoid re-signalling if the flag is set.

some-mthfka replied:
Not sure how to use `handler-bind` to accomplish this, fixed the issue in a different manner for now.

Shinmera commented on test.lisp:
```diff
@@ -123,6 +125,12 @@
     name))
 
+(defun remove-all-tests-in-package (&optional (package *package*))
+  "Remove all tests from the current PACKAGE (current by default)."
```
Docstrings belong in `documentation.lisp`, not in source forms.

Shinmera commented on test.lisp:
```diff
@@ -123,6 +125,12 @@
 
+(defun remove-all-tests-in-package (&optional (package *package*))
+  "Remove all tests from the current PACKAGE (current by default)."
+  (let ((*package* (find-package package))) ; didn't work without this
```
Instead of binding `*package*`, pass the package as the second argument to `remove-test`.

Shinmera commented on test.lisp:
```diff
@@ -161,6 +169,29 @@
+the results: if there are failures, return the status, the number of failures
+and the first failed expression; otherwise, return just the status and the
+report.  Useful for interactivity: define and run the test at once, see the
+place that needs fixed right away.  Compilation errors are not muffled."
```
Please follow the same documentation style as used in other docstrings.

Shinmera commented on test.lisp:
```diff
@@ -161,6 +169,29 @@
+  `(progn
+     (define-test ,name ,@args-and-body)
+     (eval-when (:execute)
+       (flet ((failed-p (x) (equal :FAILED (status x))))
```
Use `eq` or `eql` for comparison. Don't scream-case the keyword.

Shinmera commented on test.lisp:
```diff
@@ -161,6 +169,29 @@
+                (results (results report))
+                (status (status report)))
+           (cond ((member status '(:PASSED :SKIPPED))
+                  (list status report))
```
Returning multiple values instead of a list seems more appropriate. The first value should still always be the test's name, too.

Shinmera reviewed (changes requested):
Hi, thanks for the PR. There's a few style changes that need to be made before this can be merged.

some-mthfka commented:
Thanks for the review, I have added a `:report 'interactive` version of define-test+run and fixed the issues you mentioned. Didn't use the `handler-bind`, though, please see my comment on it.

some-mthfka commented:
@Shinmera, please let me know if there's smth else I can fix.

Shinmera commented:
Hi, sorry for being so slow to respond.

Shinmera commented on documentation.lisp:
```diff
@@ -685,6 +693,34 @@ See FIND-TEST
+    "Contains a function which takes a list of four values as per
+DEFINE-TEST+RUN and, by default, returns all of them as values (with
+VALUES-LIST).  Its purpose is to easily let you customize the feedback from
+DEFINE-TEST+RUN to your own liking.")
```
This seems too specific to include to me. A user could instead just define a new macro around `define-test+run` and use that instead to customise the results.

some-mthfka replied:
Indeed. But the user-defined expansion would have to be smth like:

```
  (defmacro my-define-test+run (name &body body)
    `(progn (eval-when (:compile-toplevel :load-toplevel)
              (parachute:define-test ,name ,@body))
            (eval-when (:execute)
              (multiple-value-bind (name report status result)
                  (parachute:define-test+run ,name ,@body)
                ;; ...
                ))))
```

so the snippet would probably best be included in the docstring. With the function, it's just:

```
(setf parachute:*postprocessor-for-define-test+run*
  (lambda (v) (destructuring-bind (name report status result) v
                ...)))
```

Yet another way is to hide the variable and expose this instead:

```
  (defmacro def-result-for-define-test+run (&body body) ; untested
    `(setf *postprocessor-for-define-test+run*
       (lambda (x) (destructuring-bind (name report status result) x
                     (declare (ignorable name report status result))
                     ,@body))))
```

What do you think?

Shinmera replied:
I suppose in order to get the best answer to this I need to understand what exactly you're using the transform for, as I struggle to find good uses for it myself. Is it just aesthetics? Are you processing the results somehow?

some-mthfka replied:
Yes, the aesthetics. So, as of now, the return result for define-test+run looks like this:

```
=> SOME-TEST-0, #<PARACHUTE:PLAIN 7, PASSED results>, :PASSED, NIL
=> SOME-TEST-0, #<PARACHUTE:PLAIN 8, FAILED results>, :FAILED, (1 (IS EQL 0 1))
```

This is a bit noisy for me, so I prefer this:

```
=> #<PARACHUTE:PLAIN 7, PASSED results>
=> (:FAILED 1), (IS EQL 0 1)
```

// the 1 in `(:FAILED 1)` shows that only one test failed

I think that could be a decent default too, btw.

Shinmera replied:
Ok. I really don't want to include a mechanism like this just for aesthetics.

I'm not sure how to feel about variable return values, but I suppose having this just return something like
```
:FAILED, (IS EQL 0 1), (IS EQL 1 2), ...
:PASSED
```
Would be fine.

some-mthfka replied:
Not a problem, please see the update. One thing: since there are implementation-defined limits on the number of multiple values (with the standard guaranteeing only 20) so I simply used a list for the failed clauses.

Shinmera replied:
Ah, good catch. That works for me!

Shinmera commented:
Thanks for the improvements!

some-mthfka commented:
My pleasure, and thanks for the review!
